### PR TITLE
[DRAFT] feat(medusa): Authentication Strategies

### PR DESCRIPTION
**Pull request by adrien2p**
### What
Add support for authentication strategies, 

The idea with that draft write up is to provide and `AbstractAuthStrategy` class that implements the interface `IAuthenticationStrategy`

The strategy that extend this abstract class must provide at least 4 required methods
- `shouldUseStrategy(req: Request, scope: "admin" | "store"): Promise<boolean>` which delegate the way to select the right strategy to the user so that way a user can provide multiple auth strategies in his store or admin (could use the request headers for example `X-auth-strategy` or something else depending on his requirements) which gives some flexibility. Since the user can handle that we don't have to enforce and care about it ourselves.
- `authenticate(req: Request, res: Response): Promise<void>` which will be call to handle the authentication on the `POST /{admin|store}/auth` route
- `unAuthenticate(req: Request, res: Response): Promise<void>` which will be call to handle the logout on the `DELETE /{admin|store}/auth` route
- `validate(req: Request, res: Response, next: NextFunction): Promise<void>` which will be call to handle the validation through medusa auth-middleware or medusa auth-customer middleware

and an optional static method can be implemented 
- `beforeInit(app: Express, container: MedusaContainer, options?: Record<string, unknown>): Promise<void>` that allow the user to be able to attach additional needed configuration on the underlying app

**Strategies loading**
The auth strategies are loaded after the express loaded has been executed, before a strategy is registered, the `beforeInit` method of the strategy constructor will be summon if provided

**Behaviour of the strategy selection**
Upon an incoming request, the authenticated routes will call the `authService` that will take the request and the scope and will take care of finding the right strategy to be used in the actual context.
If no strategies have been found then depending on the scope it will automatically fallback to either the `AdminDefaultAuthenticationStrategy` or `StoreDefaultAuthenticationStrategy`

**Authentication strategies with additional routes**
With the actual proposed approach, the user still have the default approach of adding new routes to handle for example a token renewal and can resolve his strategy by calling the `authService` just as we do in the core and then upon the resolution can call a custom method that he would have implemented first.

**Additional information**
- The reset password system still work as usual, bqsed on a jwt token and does not need any modifcation at the moment
- The invite system still work and does not need any modification at the moment, it is based on a jwt token and require the configuration to provide a secret.

### TODO
Check if some plugins must be updated as well:
- wish list plugin must be updated to use the appropriate auth strategy

**Demo**

https://user-images.githubusercontent.com/25098370/175527822-97fa866b-333c-4c1f-a8ee-354c7d85c57f.mov

### CODE REVIEW

**Comment by adrien2p** [comment]
Some work rest to be achieve like
- Create custom that automatically log in the customer 
- Config validation such as auth specific `JWT_SECRET` that should be delegated to the auth strategies
- Invitees that use also the Jwt token

File: packages/medusa/src/interfaces/authentication-strategy.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    app: Express,
+    container: MedusaContainer,
+    options: Record<string, unknown>
+  ): Promise<void> {
+    return Promise.resolve()
+  }
+
+  abstract shouldUseStrategy(
+    req: Request,
+    scope: "admin" | "store"
```
Comment (pKorsholm) [inline packages/medusa/src/interfaces/authentication-strategy.ts:33]: should we enum this?

File: packages/medusa/src/strategies/admin-authentication.ts [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+      process.env.NODE_ENV || ""
+    )
+    const errorHandler = isProduction
+      ? (msg: string): never => {
+          throw new Error(msg)
+        }
+      : console.log
+
+    const jwt_secret =
+      configModule?.projectConfig?.jwt_secret ?? process.env.JWT_SECRET
```
Comment (pKorsholm) [inline packages/medusa/src/strategies/admin-authentication.ts:164]: you added this to the config loader a couple of months back, we can just to `configmodule.projectconfig.jwt_secret`

**Review by pKorsholm — COMMENTED** [review]
I love this, I think we should do some more testing of this, maybe a PoC or two in order to really verify it's working with both admin and store authentication. 
Really nice!

**Comment by pKorsholm** [comment]
we should also consider `reset-password`

**Comment by adrien2p** [comment]
> we should also consider `reset-password`

The idea is that the default reset-password is part of the core, but if a user create a strategy that require a reset password, renew token etc he can create the custom routes that comes with his strategy and manage add the necessary methods as part of his strategy. The thing is that we can't provide all the routes and we don't know how the user want his custom end point to be handled. does it make sense

File: packages/medusa/src/services/auth.ts [hunk 3]
```diff
@@ -182,6 +199,46 @@ class AuthService extends TransactionBaseService<AuthService> {
...
+        scope === "admin"
+          ? this.strategyResolverService_.resolveAuthByType(
+              "core-admin-default-auth"
+            )
+          : this.strategyResolverService_.resolveAuthByType(
+              "core-store-default-auth"
+            )
+    }
+
+    return authStrategy
```
[the remark this answers is not in the document]
Comment (adrien2p) [reply packages/medusa/src/services/auth.ts:240]: could you provide an example of your idea?

File: packages/medusa/src/strategies/admin-authentication.ts (hunk 2 above)
Comment (adrien2p) [reply packages/medusa/src/strategies/admin-authentication.ts:164]: totaly 🚀

**Comment by olivermrbl** [comment]
Closing this in favor of a separate plugin

**Comment by adrien2p** [comment]
https://github.com/adrien2p/medusa-plugins For reference, this does the job and answer the needs

**Pull request closed without merging**